datashader/datatypes.py

```diff
@@ -0,0 +1,364 @@
+            boolean ndarray the same length as the ragged array where values
+            of True represent missing/NA values.
+        """
+        return self.mask
+
+    def take(self, indices, allow_fill=False, fill_value=None):
+        """
+        Take elements from an array.
+
+        Parameters
+        ----------
+        indices : sequence of integers
+            Indices to be taken.
+        allow_fill : bool, default False
+            How to handle negative values in `indices`.
+
+            * False: negative values in `indices` indicate positional indices
+              from the right (the default). This is similar to
+              :func:`numpy.take`.
+
+            * True: negative values in `indices` indicate
+              missing values. These values are set to `fill_value`. Any other
+              other negative values raise a ``ValueError``.
+
+        fill_value : any, default None
+            Fill value to use for NA-indices when `allow_fill` is True.
+
+        Returns
+        -------
+        RaggedArray
+
+        Raises
+        ------
+        IndexError
+            When the indices are out of bounds for the array.
+        """
+        if allow_fill:
+            sequence = [self[i] if i >= 0 else fill_value
+                        for i in indices]
+        else:
+            sequence = [self[i] for i in indices]
+
+        return RaggedArray(sequence, dtype=self.flat_array.dtype)
+
+    def copy(self, deep=False):
+        """
+        Return a copy of the array.
+
+        Parameters
+        ----------
+        deep : bool, default False
+            Also copy the underlying data backing this array.
+
+        Returns
+        -------
+        RaggedArray
+        """
+        data = dict(
+            mask=self.mask,
+            flat_array=self.flat_array,
+            start_indices=self.start_indices)
+
+        if deep:
+            # Copy underlying numpy arrays
+            data = {k: v.copy() for k, v in data.items()}
+
+        return RaggedArray(data)
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        """
+        Concatenate multiple RaggedArray instances
+
+        Parameters
+        ----------
+        to_concat : list of RaggedArray
+
+        Returns
+        -------
+        RaggedArray
+        """
+        # concat masks
+        mask = np.hstack(ra.mask for ra in to_concat)
+
+        # concat flat_arrays
+        flat_array = np.hstack(ra.flat_array for ra in to_concat)
+
+        # offset and concat start_indices
+        offsets = np.hstack([
+            [0],
+            np.cumsum([len(ra.flat_array) for ra in to_concat[:-1]])])
+
+        start_indices = np.hstack([ra.start_indices + offset
+                                   for offset, ra in zip(offsets, to_concat)])
+
+        return RaggedArray(dict(
+            mask=mask, flat_array=flat_array, start_indices=start_indices))
+
+    @property
+    def dtype(self):
+        return RaggedDtype()
```

Comment:
Maybe not a big deal, but it's a bit strange for the `id` of a `.dtype` to change on each call. Probably best to set a `._dtype` in the `__init__` and return `self._dtype` here.